src/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -0,0 +1,430 @@
+ * that future props are likely to be added. Migrations support this
+ * scenario out of the box.
+ */
+export interface SavedObjectDoc {
+  attributes: object;
+  id: string;
+  type: string;
+  migrationVersion?: MigrationVersion;
+
+  [rootProp: string]: any;
+}
+
+export type TransformFn = (doc: SavedObjectDoc) => SavedObjectDoc;
+
+type ValidateDoc = (doc: SavedObjectDoc) => void;
+
+interface MigrationDefinition {
+  [type: string]: { [version: string]: TransformFn };
+}
+
+interface Opts {
+  kibanaVersion: string;
+  migrations: MigrationDefinition;
+  validateDoc: ValidateDoc;
+}
+
+interface ActiveMigrations {
+  [type: string]: {
+    latestVersion: string;
+    transforms: Array<{
+      version: string;
+      transform: TransformFn;
+    }>;
+  };
+}
+
+/**
+ * Manages migration of individual documents.
+ */
+export interface VersionedTransformer {
+  migrationVersion: MigrationVersion;
+  migrate: TransformFn;
+}
+
+/**
+ * A concrete implementation of the VersionedTransformer interface.
+ */
+export class DocumentMigrator implements VersionedTransformer {
+  private migrations: ActiveMigrations;
+  private transformDoc: TransformFn;
+
+  /**
+   * Creates an instance of DocumentMigrator.
+   *
+   * @param {Opts} opts
+   * @prop {string} kibanaVersion - The current version of Kibana
+   * @prop {MigrationDefinition} migrations - The migrations that will be used to migrate documents
+   * @prop {ValidateDoc} validateDoc - A function which, given a document throws an error if it is
+   *   not up to date. This is used to ensure we don't let unmigrated documents slip through.
+   * @memberof DocumentMigrator
+   */
+  constructor(opts: Opts) {
+    validateMigrationDefinition(opts.migrations);
+
+    this.migrations = buildActiveMigrations(opts.migrations);
+    this.transformDoc = buildDocumentTransform({
+      kibanaVersion: opts.kibanaVersion,
+      migrations: this.migrations,
+      validateDoc: opts.validateDoc,
+    });
+  }
+
+  /**
+   * Gets the latest version of each migratable property.
+   *
+   * @readonly
+   * @type {MigrationVersion}
+   * @memberof DocumentMigrator
+   */
+  public get migrationVersion(): MigrationVersion {
+    return _.mapValues(this.migrations, ({ latestVersion }) => latestVersion);
+  }
+
+  /**
+   * Migrates a document to the latest version.
+   *
+   * @param {SavedObjectDoc} doc
+   * @returns {SavedObjectDoc}
+   * @memberof DocumentMigrator
+   */
+  public migrate = (doc: SavedObjectDoc): SavedObjectDoc => {
+    return this.transformDoc(doc);
+  };
+}
+
+/**
+ * Basic validation that the migraiton definition matches our expectations. We can't
+ * rely on TypeScript here, as the caller may be JavaScript / ClojureScript / any compile-to-js
+ * language. So, this is just to provide a little developer-friendly error messaging. Joi was
+ * giving weird errors, so we're just doing manual validation.
+ */
+function validateMigrationDefinition(migrations: MigrationDefinition) {
```

Comment:
I am not seeing any tests which cover the validation here or in `wrapWithTry`.

Reply:
Hm. Yeah. Validation could use more coverage; there's only one test.

The wrapWithTry is covered by the various error scenario tests, such as `decorates transform errors with details about what doc and transform failed`.